File: rust/agama-utils/src/api/issue.rs
// Copyright (c) [2025] SUSE LLC
//
// All Rights Reserved.
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 2 of the License, or (at your option)
// any later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
// more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, contact SUSE LLC.
//
// To contact SUSE LLC about this file by physical or electronic mail, you may
// find current contact information at www.suse.com.

use crate::api::scope::Scope;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Clone, Debug, Deserialize, Serialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct IssueWithScope {
    pub scope: Scope,
    #[serde(flatten)]
    pub issue: Issue,
}

pub type IssueMap = HashMap<Scope, Vec<Issue>>;

// NOTE: in order to compare two issues, it should be enough to compare the description
// and the details.
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq, Hash, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct Issue {
    pub class: String,
    pub description: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub details: Option<String>,
}

impl Issue {
    /// Creates a new issue.
    pub fn new(class: &str, description: &str) -> Self {
        Self {
            description: description.to_string(),
            class: class.to_string(),
            details: None,
        }
    }

    /// Sets the details for the issue.
    pub fn with_details(mut self, details: &str) -> Self {
        self.details = Some(details.to_string());
        self
    }
}

File: web/src/hooks/api/issue.ts
/*
 * Copyright (c) [2025] SUSE LLC
 *
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, contact SUSE LLC.
 *
 * To contact SUSE LLC about this file by physical or electronic mail, you may
 * find current contact information at www.suse.com.
 */

import React, { useCallback } from "react";
import { useSuspenseQuery, useQueryClient } from "@tanstack/react-query";
import { getIssues } from "~/api";
import { useInstallerClient } from "~/context/installer";
import { Issue, Scope } from "~/api/issue";

const issuesQuery = {
  queryKey: ["issues"],
  queryFn: getIssues,
};

function useIssues(scope?: Scope): Issue[] {
  const { data } = useSuspenseQuery({
    ...issuesQuery,
    select: useCallback(
      (data: Issue[]): Issue[] => (scope ? data.filter((i: Issue) => i.scope === scope) : data),
      [scope],
    ),
  });
  return data;
}

const useIssuesChanges = () => {
  const queryClient = useQueryClient();
  const client = useInstallerClient();

  React.useEffect(() => {
    if (!client) return;

    return client.onEvent((event) => {
      if (event.type === "IssuesChanged") {
        queryClient.invalidateQueries({ queryKey: ["issues"] });
      }
    });
  }, [client, queryClient]);
};

export { issuesQuery, useIssues, useIssuesChanges };
